modules/minimap2/1.0/minimap2.smk

```diff
@@ -0,0 +1,168 @@
+)
+
+#include: "../../utils/2.1/utils.smk"
+
+# Define rules to be run locally when using a compute cluster
+localrules:
+    _minimap2_input_fastq,
+    _minimap2_symlink_bam,
+    _minimap2_output_bam,
+    _minimap2_all,
+
+sample_ids_minimap2 = list(CFG['samples']['sample_id'])
+
+##### RULES #####
+
+
+# Symlinks the input files into the module results directory (under '00-inputs/')
+rule _minimap2_input_fastq:
+    input:
+        fastq = CFG["inputs"]["sample_fastq"]
+    output:
+        fastq = CFG["dirs"]["inputs"] + "fastq/{seq_type}/{sample_id}.fastq.gz"
+    run:
+        op.absolute_symlink(input.fastq, output.fastq)
+
+
+rule _minimap2_run:
+    input:
+        fastq = str(rules._minimap2_input_fastq.output.fastq),
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa")
+    output:
+        sam = pipe(CFG["dirs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}_out.sam")
+    log:
+        stdout = CFG["logs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}/minimap2.stdout.log",
+        stderr = CFG["logs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}/minimap2.stderr.log"
+    params:
+        opts = CFG["options"]["minimap2"]
+    conda:
+        CFG["conda_envs"]["minimap2"]
+    threads:
+        CFG["threads"]["minimap2"]
+    resources:
+        **CFG["resources"]["minimap2"]
+    shell:
+        op.as_one_line("""
+        minimap2 {params.opts}
+        -t {threads} 
+        {input.fasta} 
+        {input.fastq} 
+        -o {output.sam} 
+        > {log.stdout} 
+        2> {log.stderr}
+        """)
+
+
+rule _minimap2_samtools:
+    input:
+        sam = str(rules._minimap2_run.output.sam)
+    output:
+        bam = CFG["dirs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}_out.bam",
+        complete = touch(CFG["dirs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}_out.bam.complete")
+    log:
+        stderr = CFG["logs"]["minimap2"] + "{seq_type}--{genome_build}/{sample_id}/samtools.stderr.log"
+    params:
+        opts = CFG["options"]["samtools"]
+    conda:
+        CFG["conda_envs"]["samtools"]
+    threads:
+        CFG["threads"]["samtools"]
+    resources:
+        **CFG["resources"]["samtools"]
+    shell:
+        op.as_one_line("""
+        samtools view {params.opts}
+        {input.sam} > {output.bam} 
+        2> {log.stderr}
+        """)
+
+
+rule _minimap2_symlink_bam:
+    input:
+        bam = str(rules._minimap2_samtools.output.bam)
+    output:
+        bam = CFG["dirs"]["sort_bam"] + "{seq_type}--{genome_build}/{sample_id}.bam"
+    wildcard_constraints: 
+        sample_id = "|".join(sample_ids_minimap2)
+    run:
+        op.absolute_symlink(input.bam, output.bam)
+
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _minimap2_output_bam:
+    input:
+        bam = CFG["dirs"]["sort_bam"] + "{seq_type}--{genome_build}/{sample_id}.sort.bam",
```

Comment:
Is this supposed to be created by the utils module? It is commented above so is not imported. Where does this come from then?

Reply:
Yes it is! The utils module is imported by the wrapper snakefile instead of in the module